File: java-to-go-projects/05-port-scanner/Scanner.java
package scanner;

import java.io.IOException;
import java.net.*;
import java.util.*;
import java.util.concurrent.*;

/**
 * Concurrent TCP port scanner.
 *
 * Scans a range of ports using a fixed thread pool.
 * Each port gets its own task with a configurable timeout.
 */
public class Scanner {

    private final String host;
    private final int timeoutMs;
    private final int poolSize;

    public Scanner(String host, int timeoutMs, int poolSize) {
        this.host = host;
        this.timeoutMs = timeoutMs;
        this.poolSize = poolSize;
    }

    /**
     * Scan a single port. Returns a ScanResult.
     */
    public ScanResult scanPort(int port) {
        long start = System.nanoTime();
        try (Socket socket = new Socket()) {
            socket.connect(new InetSocketAddress(host, port), timeoutMs);
            long elapsed = (System.nanoTime() - start) / 1_000_000;
            return new ScanResult(host, port, true, elapsed);
        } catch (IOException e) {
            long elapsed = (System.nanoTime() - start) / 1_000_000;
            return new ScanResult(host, port, false, elapsed);
        }
    }

    /**
     * Scan a range of ports concurrently.
     * Returns only open ports, sorted by port number.
     */
    public List<ScanResult> scanRange(int startPort, int endPort) {
        ExecutorService pool = Executors.newFixedThreadPool(poolSize);
        List<Future<ScanResult>> futures = new ArrayList<>();

        for (int port = startPort; port <= endPort; port++) {
            final int p = port;
            futures.add(pool.submit(() -> scanPort(p)));
        }

        List<ScanResult> results = new ArrayList<>();
        for (Future<ScanResult> f : futures) {
            try {
                ScanResult result = f.get();
                if (result.open()) {
                    results.add(result);
                }
            } catch (Exception e) {
                // skip failed scans
            }
        }

        pool.shutdown();
        results.sort(Comparator.comparingInt(ScanResult::port));
        return results;
    }
}

File: java-to-go-projects/05-port-scanner/rust/src/scanner.rs
use tokio::{sync::Semaphore, task::JoinSet};

use crate::result::ScanResult;
use std::{net::SocketAddr, sync::Arc};

/// Scan a single port on the given host. Returns a ScanResult.
pub async fn scan_port(host: &str, port: u16, timeout_ms: u64) -> ScanResult {
    let start_time = std::time::Instant::now();
    let addr = SocketAddr::new(host.parse().expect("Invalid host"), port);
    // connect with tokio
    let open =tokio::time::timeout(
        std::time::Duration::from_millis(timeout_ms), 
        tokio::net::TcpStream::connect(addr)
    )
    .await
    .map(|res| res.is_ok())
    .unwrap_or(false);

    let latency_ms = start_time.elapsed().as_millis() as u64;

        
    ScanResult {
        host: host.to_string(),
        port,
        open,
        latency_ms,
    }

}

/// Scan a range of ports concurrently using tokio::spawn + JoinSet.
pub async fn scan_range(host: &str, start: u16, end: u16, concurrency: usize, timeout_ms: u64) -> Vec<ScanResult> {


    
    let mut join_set = JoinSet::new();
    let semaphore = Arc::new(Semaphore::new(concurrency));
    
    for port in start..=end {
        let host = host.to_string();
        let semaphore = semaphore.clone();
        join_set.spawn(async move {
            let _semaphore_permit = semaphore.acquire().await.unwrap();
            scan_port(&host, port, timeout_ms).await
        });
    }
    let mut results = Vec::new();
    while let Some(res) = join_set.join_next().await {
        if let Ok(scan_result) = res {
            results.push(scan_result);
        }
    }
    results
    
}
